### PR TITLE
Mephisto QA data collection

### PR DESCRIPTION
**Pull request by vaibhavad**
**Description**
Example of QA Data collection task using Mephisto

**Testing steps**
```
cd webapp
npm install; npm run dev
cd ..
python run.py
```

**Discussion**
There are two major points where this can be improved - 
1. Dataloader - How to seamlessly integrate the QA data loader to re-use existing ParlAI tasks? The two requirements I can think of are - (1) User should be able to specify task from config, (2) There should be a wrapper on top of original ParlAI task to modify small things (SQuAD returns both Passage and Question, the task may require only passage)
2. Passage highlighting - Many QA data collection tasks require evidence from the passage. It can be made configurable. There should be a div on top of Worker input to show the span selected from the passage. This will require modifications in the `bootstrap-chat` package.

@EricMichaelSmith Please review it and suggest changes.

### CODE REVIEW

File: parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py:94]: This method and `.shutdown()` can be removed to make use of the superclass's methods, which are the same here

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/run.py:25]: I think you can move this to a `util.py` module to keep `run.py` relatively uncluttered - also, the more ParlAI-standard way to write such a wrapper teacher would be to subclass `parlai.tasks.wrapper.agents.AbstractWrapperTeacher` and then define a custom `._edit_action()` method in which to modify the `'text'` field of the action

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/run.py:50]: See notes above for how to simplify this class - I think only 2 params here are truly necessary, `datatype` and `task`

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/run.py:76]: I'd remove this param entirely, because the QA task is limited to 2 "turns" (a question turn and an answer turn)

File: parlai/crowdsourcing/tasks/qa_data_collection/task_description.html
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/task_description.html:5]: Hmm - I'd probably remove the "(or perhaps more turns" line, because more than 2 turns doesn't appear to be supported by the code at the moment

File: parlai/crowdsourcing/tasks/qa_data_collection/task_description.html
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/qa_data_collection/task_description.html:6]: It looks like we don't have a form here, so I'd remove this line too

**Comment by EricMichaelSmith** [comment]
Wow, I love this - thanks @vaibhavad for adding this! It'll be a great addition to our set of Mephisto crowdsourcing tasks. I've added a bunch of comments, mostly relating to standard ParlAI and `parlai.crowdsourcing` conventions, but it looks like the major components are all here. To address your points below:
1. Yeah, there's a standard ParlAI way to load data - see my comments on this, but happy to explain further if you'd like. 
2. I think showing evidence would be a great add-on here: I'll defer to @JackUrb and @pringshia for advice on configuring the front-end code for this

Great job!

**Review by EricMichaelSmith — COMMENTED** [review]
Wow, I love this - thanks @vaibhavad for adding this! It'll be a great addition to our set of Mephisto crowdsourcing tasks. I've added a bunch of comments, mostly relating to standard ParlAI and parlai.crowdsourcing conventions, but it looks like the major components are all here. To address your points below:

Yeah, there's a standard ParlAI way to load data - see my comments on this, but happy to explain further if you'd like.
I think showing evidence would be a great add-on here: I'll defer to @JackUrb and @pringshia for advice on configuring the front-end code for this
Great job!

**Comment by JackUrb** [comment]
Hi @vaibhavad, thanks for bringing this together! Definitely agreed that adding more context to a task such that it's not all inline in the chat would be great. In `bootstrap-chat` we support overriding with custom components within a chat application by providing a file containing those custom components. See how `renderSidePane` is updated in the `custom_simple` version of the ParlAI Chat demo [here](https://github.com/facebookresearch/Mephisto/blob/fe2c5b65a07c35c7b9df1a79718fcba79809b22f/examples/parlai_chat_task_demo/custom_simple/main.js#L48). We even have setups where the context is always updating, like what's present in the PR for multimodal chat [here](https://github.com/facebookresearch/Mephisto/pull/328/files#diff-76d1b04ec07fc44067d32500e6d54a7f3c206df5cf32c2180b81834a68807770R22-R77).

While it'd be a great enhancement for sure, I don't think it should stop this from getting merged as is, as this was the level of usability we had for the original task. Can always open a follow-up PR if you want to contribute more!

File: parlai/crowdsourcing/tasks/qa_data_collection/conf/base.yaml
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/conf/base.yaml]: Makes sense. I have shifted everything to `example.py` (like in other task examples), as there were not many params remaining in `base.yaml`. Hope that's okay

File: parlai/crowdsourcing/tasks/qa_data_collection/conf/onboarding_example.yaml
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/conf/onboarding_example.yaml]: Removed `onboarding_example.yaml. Should I also remove `make_onboarding_world`, `validate_onboarding`, and `QADataCollectionOnboardWorld` from `worlds.py`?

File: parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py]: Changed it to `dataloader` to `teacher` everywhere

File: parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py]: Sorry, it got copied from another task, where we were having 3 turn conversation - question, answer, evidence

The current implementation asks question and answer in the same parley call i.e. parley is called only once. Can this lead to any unintended consequences? If so, I can add return statements [here](https://github.com/facebookresearch/ParlAI/pull/3318/files#diff-96eafa5491b76bc19d4d317fb7a6fccb9aeeb9a179b33a1b87a70b4ade3d5fb7R72)

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Does `AbstractWrapperTeacher` and `parlai.core.worlds.create_task` have different workflows? For SQuAD, `AbstractWrapperTeacher` was asking for `datapath`, but `create_task` is working without `datapath`

File: parlai/crowdsourcing/tasks/qa_data_collection/conf/onboarding_example.yaml
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/qa_data_collection/conf/onboarding_example.yaml]: Yeah, good catch - honestly, all of those should be able to be removed as well because we're not doing onboarding

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Yeah, no need for this in QA collection. I have removed this

File: parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/qa_data_collection/demo_worlds.py]: Hmm, I don't think there should be any unintended consequences - I just checked, and in the Mephisto repo it just looks like `world.parley()` is called over and over until `world.episode_done()` returns True. However, I just checked the documentation, and the docstring for `CrowdTaskWorld.parley()` states that a parley should represent one turn of your task. So I think the cleanest thing here would be to add a `return` statement at the end of the `if not self.question:` block, yes, so that `.parley()` will always be called exactly twice

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Oh hmm, you could actually combine `AbstractWrapperTeacher` and `create_task` together, like this:
```
from parlai.agents.repeat_label.repeat_label import RepeatLabelAgent
from parlai.core.params import ParlaiParser
from parlai.core.worlds import create_task


args = """\
--task wrapper:LabelToText
--wrapper-task squad
--datatype train
"""
parser = ParlaiParser(True, False)
opt = parser.parse_args(args.split())
agent = RepeatLabelAgent(opt)
teacher = create_task(opt, agent).get_task_agent()
```
Here, `LabelToText` is a teacher that subclasses `AbstractWrapperTeacher`, but you could make a similar subclassed teacher that does the string manipulation that you want

Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Thanks! :) I was looking for something as simple as this! Made changes in 104e07a0f49587b2f37b91ab19384d4be28106b6

**Comment by vaibhavad** [comment]
Hi @JackUrb,

Thanks a lot for your comments. I'll explore overriding components in `bootstrap-chat`, and include evidence marking for QA tasks in a follow-up PR as you suggested.

File: parlai/crowdsourcing/tasks/qa_data_collection/run.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Okay great, that works! But you probably don't need both a `task` and `wrapper_task` config param in `run.py` - it looks like this wrapper teacher is specific to SQuAD, so you might as well just hardcode in that `--wrapper-task` has to be `'squad'`, maybe by adding in a `parser.set_defaults(wrapper_task='squad')` line in your subclass's `.add_cmdline_args()` method. Also, since this subclassed wrapper teacher is SQuAD-specific, it's likely more suitable to go in `parlai.tasks.squad.agents` than in `parlai.tasks.wrapper.agents`

Comment (vaibhavad) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Changes to remove `wrapper_task` are in 870690f71c9829028ddd9639532a72d281aa387c

Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/qa_data_collection/run.py]: Ah fantastic! Hmm, in the TeacherConfig, can you set the default to `"squad:SquadQATeacher"`, just so that the teacher works on the default settings? Also, super minor comment, but in the SquadQATeacher, it might be good to put the add_cmdline_args() method before the other methods for consistency, since we typically put the classmethods before the rest. Other than that I think this looks good :)

**Review by EricMichaelSmith — APPROVED** [review]
Thanks so much for all of the refactors on this! Approving now. It looks like Jack is out of the office until next week, but it seems like he agrees that the remaining improvements can be done in a follow-up PR, so it's fine with me if you merge in :)

**Comment by vaibhavad** [comment]
Thanks @EricMichaelSmith for all the constructive feedback throughout. Feels good to have my first contribution.

I don't think I have permission to merge. You or @JackUrb can merge it if no more changes are required.

**Comment by EricMichaelSmith** [comment]
> Thanks @EricMichaelSmith for all the constructive feedback throughout. Feels good to have my first contribution.
> 
> I don't think I have permission to merge. You or @JackUrb can merge it if no more changes are required.

Okay cool! It looks like some of the CI checks are failing (linting and stuff) - this is pretty minor, but when I have time today I'll fix those up before merging in (unless you want to take a crack at them first, totally up to you)

**Pull request merged**
